codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -185,125 +285,167 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
             // TODO: improve handler/stream lifecycle so that stream isn't active before handler added.
             // The stream was already made active, but ctx may have been null so it wasn't initialized.
             // https://github.com/netty/netty/issues/4942
             new ConnectionListener().onStreamActive(stream);
             upgrade.upgradeRequest().headers().setInt(
                     HttpConversionUtil.ExtensionHeaderNames.STREAM_ID.text(), Http2CodecUtil.HTTP_UPGRADE_STREAM_ID);
-            new InboundHttpToHttp2Adapter(http2Handler.connection(), http2Handler.decoder().frameListener())
+            new InboundHttpToHttp2Adapter(connection(), decoder().frameListener())
                     .channelRead(ctx, upgrade.upgradeRequest().retain());
         } finally {
             upgrade.release();
         }
     }
 
-    // Override this to signal it will never throw an exception.
-    @Override
-    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) {
-        ctx.fireExceptionCaught(cause);
-    }
-
     /**
      * Processes all {@link Http2Frame}s. {@link Http2StreamFrame}s may only originate in child
      * streams.
      */
     @Override
     public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
-        try {
-            if (msg instanceof Http2WindowUpdateFrame) {
-                Http2WindowUpdateFrame frame = (Http2WindowUpdateFrame) msg;
-                consumeBytes(frame.streamId(), frame.windowSizeIncrement(), promise);
-            } else if (msg instanceof Http2StreamFrame) {
-                writeStreamFrame((Http2StreamFrame) msg, promise);
-            } else if (msg instanceof Http2GoAwayFrame) {
-                writeGoAwayFrame((Http2GoAwayFrame) msg, promise);
-            } else {
-                throw new UnsupportedMessageTypeException(msg);
-            }
-        } finally {
+        if (msg instanceof Http2DataFrame) {
+            Http2DataFrame dataFrame = (Http2DataFrame) msg;
+            encoder().writeData(ctx, dataFrame.stream().id(), dataFrame.content(),
+                    dataFrame.padding(), dataFrame.isEndStream(), promise);
+        } else if (msg instanceof Http2HeadersFrame) {
+            writeHeadersFrame(ctx, (Http2HeadersFrame) msg, promise);
+        } else if (msg instanceof Http2WindowUpdateFrame) {
+            Http2WindowUpdateFrame frame = (Http2WindowUpdateFrame) msg;
+            writeWindowUpdate(frame.stream().id(), frame.windowSizeIncrement(), promise);
+        } else if (msg instanceof Http2ResetFrame) {
+            Http2ResetFrame rstFrame = (Http2ResetFrame) msg;
+            encoder().writeRstStream(ctx, rstFrame.stream().id(), rstFrame.errorCode(), promise);
+        } else if (msg instanceof Http2PingFrame) {
+            writePingFrame(ctx, (Http2PingFrame) msg, promise);
+        } else if (msg instanceof Http2SettingsFrame) {
+            writeSettingsFrame(ctx, (Http2SettingsFrame) msg, promise);
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)